### PR TITLE
Pass Variable into Caffe2 ops, by not checking `!is_variable()` in caffe2::Tensor constructor

### PR DESCRIPTION
**Pull request by yf225**
As part of the Variable/Tensor merge, we want to be able to pass Variables into Caffe2 without doing extra shallow copy. There are three possible approaches to this:

1. Pass Variable into Caffe2 ops, by adding `at::AutoNonVariableTypeMode` to `Run()`/`RunAsync()` methods in Caffe2 `Operator` class. (PR: https://github.com/pytorch/pytorch/pull/22284)
Problem: There are existing Caffe2 operators (e.g. `ScriptModuleOp`) that call JIT operations with Variables as input: https://github.com/pytorch/pytorch/blob/f5b3f9ecd9c21a720d90012cd4d71071b37c24df/caffe2/contrib/pytorch/script_module_op.cc#L97-L101
Since the JIT interpreter takes only Variables, in order to make the `ScriptModuleOp` work we cannot use `at::AutoNonVariableTypeMode` to treat all Variables as Tensors in Caffe2.

2. Pass Variable into Caffe2 ops, by requiring that all Variables in Caffe2 don't require grad. (PR: https://github.com/pytorch/pytorch/pull/22473)
Problem: There are existing Caffe2 operators (e.g. `ScriptModuleOp`) that operate on a ScriptModule with requires-grad parameters:
the ScriptModule is defined in:
https://github.com/pytorch/pytorch/blob/c9f41e9bc0c8296daede62ee936b9b7e32a8b658/caffe2/python/workspace_test.py#L730-L746
output from the ScriptModule's `forward` is computed in:
https://github.com/pytorch/pytorch/blob/c9f41e9bc0c8296daede62ee936b9b7e32a8b658/caffe2/contrib/pytorch/script_module_op.cc#L104-L116
Since the ScriptModule's parameter requires grad, output from the ScriptModule's `forward` also requires grad, and thus we cannot assume that the input to the `castIValueToTensor()` call above doesn't require grad, which violates our invariant that all Variables in Caffe2 don't require grad.

3. **[Chosen Approach]** Pass Variable into Caffe2 ops, by not checking `!is_variable()` in `caffe2::Tensor()` constructor. (this PR)
Why this approach works: Since Caffe2 operators do not dispatch to `VariableType`, whether the `caffe2::Tensor` has AutogradMeta should be irrelevant, and autograd history is never tracked in Caffe2. Therefore, it's safe to remove the `!is_variable()` check in `caffe2::Tensor()` constructor, since Caffe2 ops will safely work with Variables.

### CODE REVIEW

**Comment by gchanan** [comment]
> As part of the Variable/Tensor merge, we want to be able to pass Variables into Caffe2 without doing extra shallow copy.

Why do we want to do that?  Is it performance or is this a requirement because we do in-place mutations in Caffe2 ops or something?

**Comment by gchanan** [comment]
> Pass Variable into Caffe2 ops, by requiring that all Variables in Caffe2 don't require grad. (PR: #22473)
Problem: There are existing Caffe2 operators (e.g. ScriptModuleOp) that operate on a ScriptModule with non-requires-grad parameters:

Is this last statement supposed to be "requires-grad", not "non-requires-grad" ?

**Comment by gchanan** [comment]
@dzhulgakov I'm not sure I understand all the C2 constraints here -- can you take a look?

**Comment by dzhulgakov** [comment]
The only point ScriptModuleOp is used is for bridging some pretrainined models / doing preprocessing. I don't think we will connect autograd to C2 gradients at any point as it's too tricky and not that useful.

So I'd safely assume that we can treat any point crossing PT->C2 to enforce that require_grad=False (it's a stronger contract). For ScriptModuleOp specifically, I think we should just wrap its execution with `torch::NoGradGuard` and it's solve the problem you pointed above. Thus overall it looks like #22473 is pretty close to what you want.

**Comment by yf225** [comment]
@smessmer Would you like to take a look at this?

**Comment by smessmer** [comment]
Just removing the assertion should work (or at least I don't see any obvious things that break), but it might cause weird behavior for autograd because caffe2 operators don't implement autograd correctly. Calling a caffe2 op with `requires_grad=True` almost certainly doesn't do what you'd expect it to, maybe with the exception of `ScriptModuleOp`. I agree with @dzhulgakov that we should enforce `requires_grad=False` at the boundary between the frameworks.

**Comment by yf225** [comment]
This is superseded by https://github.com/pytorch/pytorch/pull/22473.
